text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,455 @@
+
+Backend-specific syntactic sugar for people that definitely only need one backend.
+
+```ts
+const pipeline = new cdkp.CodePipeline(this, 'Pipeline', {
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    ...
+  }),
+  ...backendOptions
+});
+pipeline.addCdkStage(...);
+pipeline.addCdkStage(...);
+
+pipeline.render();  // <-- can be automatically called if not called at `prepare()` time.
+```
+
+Advantages of options 2, 3 are that `render()` can be implicit if necessary, and so cannot
+be accidentally forgotten.
+
+Recommendation: we implement options 1 and 3.
+
+### Waves
+
+Deploying multiple applications in parallel is done by adding waves:
+
+```ts
+const def = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({ ... }),
+  waves: [
+    {
+      waveName: '...',
+      stages: [
+        {
+          stageName?: '...',
+          stage: new MyAppStage(...),
+          after: [...],
+        }
+      ],
+    },
+  ],
+});
+def.addCdkStage(new MyAppStage(...));
+
+const wave = def.addWave('asdf', {
+  stages: [...],
+});
+wave.addCdkStage(new MyAppStage(...));
+```
+
+### Validations
+
+Waves and Stages can be decorated with `before` and `after` actions, intended for validations
+and checks.
+
+```ts
+const def = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({ ... }),
+});
+def.addCdkStage(new MyAppStage('Beta', ...));
+
+const wave = def.addWave('Prod', {
+  before: [
+    new cdkp.ManualApprovalAction('Promote Beta to Prod'),
+  ],
+});
+
+wave.addCdkStage(new MyAppStage(...), {
+  after: [
+    new cdkp.ShellScriptAction('Validate', {
+      commands: [
+        'curl -Ssf https://mywebsite.com/ping',
+      ],
+    }),
+  ],
+});
+```
+
+### Using additional artifacts
+
+To clearly distinguish them from Stack Outputs, we will call artifacts "file sets". They
+can be consumed and produced by build actions and shell actions, and are always
+about whole directories:
+
+```ts
+const pipeline = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    additionalInputs: {
+      'subdir': cdkp.CodePipelineSource.gitHub('OWNER2/REPO2'),
+      '../side-dir': someStep,
+    },
+    commands: ['npm ci', 'npm run build'],
+    additionalOutputs: {
+      tests: cdkp.AdditionalOutput.fromDirectory('test'),
+    }
+  }),
+});
+
+pipeline.addCdkStage(stage, {
+  after: [
+    new cdkp.ShellScriptAction({
+      input: pipeline.additionalBuildOutput('tests'),
+      commands: ['node ./integ-tests'],
+      output: cdkp.AdditionalOutput.fromDirectory('./test-results'),
+    }),
+  ],
+});
+```
+
+### Using Stack Outputs
+
+Stack outputs are useful in validations:
+
+```ts
+const cdkStage = new MyAppStage(...);
+wave.addCdkStage(cdkStage, {
+  after: [
+    new cdkp.ShellScriptAction('Validate', {
+      environmentVariablesFromStackOutputs: {
+        ENDPOINT_URL: cdkp.StackOutput.fromCfnOutput(cdkStage.endpointUrl),
```

Comment:
Is it possible to use string tokens instead of a strong type and then deduce automatically if needed?

```ts
env: {
  ENDPOINT_URL: cdkStage.endpointUrl
}
```

Reply:
Mmmmmmmmmmmmaybe. We can investigate that.

Reply:
I looked into it some. There isn't actually an existing token that refers to a CfnOutput, so there's no stack.output.ref.

We can do something like StackOutputRef.fromCfnOutput(stack.output), encode that into a string and then decode it later... but that seems to be growing silly.